plugins/shadow-block-converter/src/shadow_block_converter.js

```diff
@@ -0,0 +1,194 @@
+   * @override
+   */
+  isNull() {
+    return this.oldValue === this.newValue;
+  }
+
+  /**
+   * Run a change event.
+   * @param {boolean} forward True if run forward, false if run backward (undo).
+   * @override
+   */
+  run(forward) {
+    const workspace = this.getEventWorkspace_();
+    const block = workspace.getBlockById(this.blockId);
+    if (!block) {
+      console.warn('Can\'t change non-existent block: ' + this.blockId);
+      return;
+    }
+
+    const value = forward ? this.newValue : this.oldValue;
+    block.setShadow(!!value);
+  }
+}
+
+Blockly.registry.register(
+    Blockly.registry.Type.EVENT,
+    BlockShadowChange.EVENT_TYPE,
+    BlockShadowChange);
+
+/**
+ * Add this function to your workspace as a change listener to automatically
+ * convert shadow blocks to real blocks whenever the user edits a field on the
+ * block, like this:
+ *
+ * workspace.addChangeListener(shadowBlockConversionChangeListener);
+ *
+ * Ideally the Blockly.Field.prototype.setValue method should handle this logic,
+ * but for the purposes of this plugin it can be a workspace change listener.
+ *
+ * @param {Blockly.Events.Abstract} event An event broadcast by the workspace.
+ */
+export function shadowBlockConversionChangeListener(event) {
+  // Auto-converting shadow blocks to real blocks should happen in response to
+  // new user action events (which get recorded as undo events) but not when
+  // undoing or redoing events (which do not get recorded again).
+  if (!event.recordUndo) {
+    return;
+  }
+
+  // Auto-converting shadow blocks to real blocks should happen in response to
+  // editing a field value, which is recorded as Blockly.Events.BLOCK_CHANGE.
+  if (event.type != Blockly.Events.BLOCK_CHANGE) {
+    return;
+  }
+
+  const workspace = Blockly.Workspace.getById(event.workspaceId);
+  if (!workspace) {
+    return;
+  }
+  const block = workspace.getBlockById(event.blockId);
+  if (!block) {
+    return;
+  }
+
+  // Blocks that are already real blocks can be ignored.
+  if (!block.isShadow()) {
+    return;
+  }
+
+  // Remember the current event group so that it can be resumed below.
+  const currentGroup = Blockly.Events.getGroup();
+
+  if (event.group) {
+    // Temporarily use the same group as the initiating event so that
+    // the shadow events get grouped with it for undo purposes.
+    Blockly.Events.setGroup(event.group);
+  } else {
+    // The initiating event wasn't part of any named group, so the shadow events
+    // can't be grouped with it, but at least they can be grouped with each
+    // other. This happens when editing e.g. boolean fields.
+    Blockly.Events.setGroup(true);
+
+    // HACK: Retroactively overwrite the initiating event's group property in
+    // order to include it in the same group as the shadow events anyway. It
+    // wasn't previously grouped with any other events, but now it can be
+    // grouped with the new events.
+    event.group = Blockly.Events.getGroup();
```

Comment:
Oh. Yeah, this is pretty gross. It will work if no one else is messing around with the event stream, but it's definitely a hack.

Reply:
Reverted. I'll file an issue on blockly about why editing boolean fields results in ungroupable change events.